### PR TITLE
Add new resource `random_bytes` to generate an array of random bytes

### PR DESCRIPTION
**Pull request by Socolin**
Intended to be used as key or secret

Fix https://github.com/hashicorp/terraform-provider-random/issues/63

### CODE REVIEW

**Comment by Socolin** [comment]
@bendbennett Hello, is there any chance to see this PR merge (or if it's not something you want, rejected) soon ? I need this I would like to know if at least it's something you are considering or if I should instead create my own provider ?

**Comment by bendbennett** [comment]
@Socolin apologies for delay in responding on this. I'm afraid that I don't have an ETA at this time but I will endeavour to get some feedback to you.

**Comment by dnlopes** [comment]
I'm also waiting badly for this PR to get merged. Have a comment in my code base pointing here 😄

File: CHANGELOG.md
Comment (bendbennett) [inline CHANGELOG.md:6]:
```suggestion
## 3.5.0 (unreleased)

NEW FEATURES:

* resource/random_bytes: Generates an array of random bytes that is intended to be used as key or secret ([#272](https://github.com/hashicorp/terraform-provider-random/pull/272))

```

**Review by bendbennett — COMMENTED** [review]
Thanks for the contribution.

Could you rebase on `main`?
Looks like this PR is using `v0.9.0` of `terraform-plugin-framework`.

**Comment by sebhaub** [comment]
nice - this PR is exactly what i currently miss in the random provider. 
@bendbennett is there something missing / preventing to merge this ?

**Comment by jmgilman** [comment]
I'd be willing to help out if it means getting this merged sooner. A huge need for us, thanks!

**Comment by Socolin** [comment]
@bendbennett Do you have an ETA on this ?

**Comment by bookshelfdave** [comment]
We hope to release this in the next few weeks, it's queued behind some other work that we have planned.

File: CHANGELOG.md
Comment (bendbennett) [inline CHANGELOG.md:6]:
```suggestion
```

We have recently switched to using [Changie](https://changie.dev/) for changelog and version management.

Could you remove this entry from CHANGELOG.md and once Changie is [installed](https://changie.dev/guide/installation/) run the following:

```shell
changie new
? Kind: Features
Body: resource/random_bytes: New resource that generates an array of random bytes intended to be used as key or secret ([#63](https://github.com/hashicorp/terraform-provider-random/issues/63)).
Issue/PR Number: 272
```

File: internal/provider/resource_bytes.go
Comment (bendbennett) [inline internal/provider/resource_bytes.go:124]: As this is the first version of the schema for the `bytes` resource, we could use `Version: 0`.

File: internal/provider/resource_bytes.go
Comment (bendbennett) [inline internal/provider/resource_bytes.go:122]: If `Version: 0` is used for the schema the function name could be `bytesSchemaV0()`.

**Review by bendbennett — COMMENTED** [review]
@Socolin thank you for your contribution.

Would you be happy to add a couple more tests to verify:

- Behaviour of validator on `length` attribute is as expected.
- Altering `length` forces a replacement.

File: docs/resources/bytes.md
Comment (maaarghk) [inline docs/resources/bytes.md:8]: should be `random_bytes`

File: docs/resources/bytes.md
Comment (maaarghk) [inline docs/resources/bytes.md:50]: mentioning this is base64 is probably justified here

File: internal/provider/resource_bytes.go
Comment (maaarghk) [inline internal/provider/resource_bytes.go:163]: at a cursory glance I don't think this is accurate, "id" appears to always be set to "none". Other providers have this described as "A static value used internally by Terraform, this should not be referenced in configurations."

Should `result` be added to actually return the generated bytes in string format? probably a question for @bendbennett - not sure how the state file is expected to behave with binary data in it - or whether it could be achieved in the `Read` function - or if anyone would ever use it anyway

File: internal/provider/resource_bytes.go
Comment (maaarghk) [inline internal/provider/resource_bytes.go:103]: nit: I don't think re-encoding is required here, can just do `req.ID`

File: internal/provider/resource_bytes_test.go
Comment (maaarghk) [inline internal/provider/resource_bytes_test.go:29]: nit: the above comments should refer to bytes rather than password I suppose

File: internal/provider/resource_bytes_test.go
Comment (maaarghk) [inline internal/provider/resource_bytes_test.go:40]: I guess this should be updated to have `random_bytes.basic` for the id and to return `result_base64`?

**Review by maaarghk — CHANGES_REQUESTED** [review]

File: internal/provider/resource_bytes.go
Comment (Socolin) [reply internal/provider/resource_bytes.go:163]: I did not set `result` to avoid confusion, since the result is always encoded. But I could add it, we just need to choose if the default is base64 or hex (or other ?)

**Comment by Socolin** [comment]
Thanks for the comments, I pushed a fix for them.

**Comment by maaarghk** [comment]
> Thanks for the comments, I pushed a fix for them.

Think everything I left looks good, although maybe I obscured a request from Ben for two more tests to be added :)

**Comment by Socolin** [comment]
> > Thanks for the comments, I pushed a fix for them.
> 
> Think everything I left looks good, although maybe I obscured a request from Ben for two more tests to be added :)

Thanks I forgot about this, I'm checking this

**Comment by Socolin** [comment]
I added 2 tests. I noticed that there are many tests for `keepers` in the `random_password` resource. Should there also be tests for random_bytes? If so, could you share tests for both resources instead of copy/pasting them?

File: .changes/unreleased/FEATURES-20230213-213454.yaml
Comment (bendbennett) [inline .changes/unreleased/FEATURES-20230213-213454.yaml:3]:
```suggestion
  intended to be used as key or secret'
```

Apologies, believe I suggested including a link in the changie file. This is not needed as changie will generate a link on the basis of the specified `Issues`.

File: internal/provider/resource_bytes_test.go
Comment (bendbennett) [inline internal/provider/resource_bytes_test.go:87]:
```suggestion
```

Looks like the same value is being extracted twice.

File: internal/provider/resource_bytes.go
Comment (bendbennett) [reply internal/provider/resource_bytes.go]: As the base64 and hex encodings of `result` are available it doesn't seem like there is a need for `result` as @Socolin indicates it would need to be encoded.

**Comment by bendbennett** [comment]
> I added 2 tests. I noticed that there are many tests for `keepers` in the `random_password` resource. Should there also be tests for random_bytes? If so, could we share tests for both resources instead of copy/pasting them?

Believe some tests around `keepers` would be useful. I think code duplication for tests is ok in this case.

File: internal/provider/resource_bytes_test.go
Comment (Socolin) [reply internal/provider/resource_bytes_test.go:87]: Line was duplciated by error, I fix

**Comment by Socolin** [comment]
I added tests for `keepers`  but 2 test don't pass and I don't understand why

`TestAccResourceBytes_Keepers_Keep_NullMapToNullValue`
`TestAccResourceBytes_Keepers_Keep_EmptyMapToNullValue`

Do you know why ? @bendbennett

File: internal/provider/resource_bytes.go
Comment (bendbennett) [inline internal/provider/resource_bytes.go:83]:
```suggestion
func (r *bytesResource) Update(context.Context, resource.UpdateRequest, *resource.UpdateResponse) {
	var model bytesModelV0

	resp.Diagnostics.Append(req.Plan.Get(ctx, &model)...)

	if resp.Diagnostics.HasError() {
		return
	}

	resp.Diagnostics.Append(resp.State.Set(ctx, &model)...)
}
```

File: internal/provider/resource_bytes.go
Comment (bendbennett) [inline internal/provider/resource_bytes.go:153]:
```suggestion
				Sensitive:   true,
				PlanModifiers: []planmodifier.String{
					stringplanmodifier.UseStateForUnknown(),
				},
```

File: internal/provider/resource_bytes.go
Comment (bendbennett) [inline internal/provider/resource_bytes.go:158]:
```suggestion
				Sensitive:   true,
				PlanModifiers: []planmodifier.String{
					stringplanmodifier.UseStateForUnknown(),
				},				
```

File: internal/provider/resource_bytes.go
Comment (bendbennett) [inline internal/provider/resource_bytes.go:162]:
```suggestion
				Computed:    true,
				PlanModifiers: []planmodifier.String{
					stringplanmodifier.UseStateForUnknown(),
				},				
```

**Comment by bendbennett** [comment]
> I added tests for `keepers` but 2 test don't pass and I don't understand why
> 
> `TestAccResourceBytes_Keepers_Keep_NullMapToNullValue` `TestAccResourceBytes_Keepers_Keep_EmptyMapToNullValue`

The issue here is that there is a discrepancy between the `plan` and the `state` during the `Update`.

The `mapplanmodifiers.RequiresReplaceIfValuesNotNull()` does not mark `keepers` as `RequiresReplace` because `keepers` is updated to:
```
keepers = {
	"key" = null
}
```  

This results in a call to the `Update` function on the resource. As this function is empty, the `state` that is passed in via `resource.UpdateRequest` is unaltered in `resource.UpdateResponse`. Terraform detects that there is a mismatch between the `plan` and the `state` and indicates this through:
```
        Error: Provider produced inconsistent result after apply
        
        When applying changes to random_bytes.test, provider
        "provider[\"registry.terraform.io/hashicorp/random\"]" produced an unexpected
        new value: .keepers: was
        cty.MapVal(map[string]cty.Value{"key":cty.NullVal(cty.String)}), but now
        null.
```

I've added some comments into the PR with the required code changes.

**Comment by Socolin** [comment]
Ok thanks for the explanation, it should now be good to merge I think.

**Comment by dhermes** [comment]
> Ok thanks for the explanation, it should now be good to merge I think.

This is great to see! Cheers everyone

**Comment by Socolin** [comment]
I just saw the build failed (https://github.com/hashicorp/terraform-provider-random/actions/runs/4179417990), I think the reason is that I did not ran `make generate`, it's now done.

File: internal/provider/resource_bytes.go
Comment (bflad) [inline internal/provider/resource_bytes.go:146]: Drive-by comment: This new resource could use the regular `mapplanmodifier.RequiresReplace()` as there shouldn't be any prior state storage from a terraform-plugin-sdk based resource which would've collapsed a map of null values into a null map (the intended usage of the custom plan modifier was to prevent replacing existing resources stored with the prior SDK's invalid state). It may be worth an additional code comment here on the design decision either which way, whether to keep consistency with the other resources or not (so all `keepers` changes including adding `null` values replace the resource).

Comment (Socolin) [reply internal/provider/resource_bytes.go:146]: @bendbennett should I change this or keep it that way ?

Comment (bendbennett) [reply internal/provider/resource_bytes.go:146]: @Socolin just for the sake of consistency with the other resources, I think I'd be inclined to use `mapplanmodifier.RequiresReplace()` and add a code comment as @bflad suggests.

Comment (Socolin) [reply internal/provider/resource_bytes.go:146]: What the comment should say ?  something like ?:

`mapplanmodifier.RequiresReplace()` could be used, but for consistency with other resources we use  `mapplanmodifiers.RequiresReplaceIfValuesNotNull() ` 

or do you have any other suggestion ? @bendbennett

Comment (bendbennett) [reply internal/provider/resource_bytes.go:146]: @Socolin it might be useful to refer to the fact that _random_bytes_ was not available as a a resource on the _random_ provider until after the migration from SDKv2 to the Framework. 

Perhaps something like:
> mapplanmodifiers.RequiresReplaceIfValuesNotNull() has been used for consistency with other resources but mapplanmodifier.RequiresReplace() could have been used as there shouldn't be any prior state storage from a terraform-plugin-sdk based resource which would've collapsed a map of null values into a null map.

Comment (Socolin) [reply internal/provider/resource_bytes.go:146]: Ok, thanks.

**Comment by Socolin** [comment]
I fixed the last review comment

File: internal/provider/resource_bytes.go
Comment (bendbennett) [inline internal/provider/resource_bytes.go:81]: Believe this comment can be removed after the last set of changes.

File: internal/provider/resource_bytes.go
Comment (bendbennett) [inline internal/provider/resource_bytes.go:116]: I think perhaps we want to use `state.Keepers = types.MapNull(types.StringType)` here otherwise there will be a forced update following import if the configuration does not contain an empty `keepers` attribute. For example,

```terraform
resource "random_bytes" "basic" {
  length = 16
  keepers = {}
}
```

Could you add a test to ensure that following import when apply is run there are no planned changes?

Comment (Socolin) [reply internal/provider/resource_bytes.go]: I added a test, `types.MapNull(types.StringType)` does not seems to changes the outcome, but with the test I think it's validating this scenario

**Comment by dominicroessner** [comment]
Any update on when this will be released? We’re looking to use this also.

File: internal/provider/resource_bytes_test.go
Comment (bendbennett) [inline internal/provider/resource_bytes_test.go:74]: Using `types.MapNull(types.StringType)` rather than `types.MapValueMust(types.StringType, nil)` prevents an in-place update if the `random_bytes` resource is imported and then `terraform apply` is run.

Could you amend the test as follows:

```suggestion
				PlanOnly: true,
```

If you try running this test whilst using `types.MapValueMust(types.StringType, nil)` you can see that it fails with the following error:

```          # random_bytes.basic will be updated in-place
          ~ resource "random_bytes" "basic" {
                id            = "none"
              - keepers       = {} -> null
                # (3 unchanged attributes hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
```

File: internal/provider/resource_bytes_test.go
Comment (bendbennett) [inline internal/provider/resource_bytes_test.go:68]:
```suggestion
```

Comment (bendbennett) [reply internal/provider/resource_bytes_test.go:68]: I think you can delete `ResourceName: "random_bytes.basic",`, I don't think it's required.

Comment (Socolin) [reply internal/provider/resource_bytes_test.go]: > I think you can delete `ResourceName: "random_bytes.basic",`, I don't think it's required.

I forgot to push my commit, it's fixed

**Comment by Socolin** [comment]
I rebased the branch on `main` too

**Review by bendbennett — APPROVED** [review]

**Comment by ohemelaar** [comment]
Hello,
I've been lurking at this PR for a while now, I'm quite interested in using this. I see the PR is approved, is there a specific release schedule that's blocking the merge now?

**Comment by dominicroessner** [comment]
Any update or timeline on when this might go in would really be appreciated. Waiting for this as others. Thank you.

**Comment by Socolin** [comment]
Hello,

@bendbennett What is the next steps for this to be merged and released ?

**Comment by bendbennett** [comment]
Hi @Socolin 👋

I'd like to get an additional approval of the PR prior to merging as these changes are a significant alteration to the _random_ utility provider. The efforts of the team are currently focussed elsewhere at the moment but we will endeavour to focus some attention on this as soon as we're able.

**Comment by GerardSoleCa** [comment]
Hi @bendbennett, another interested party here. I'd really appreciate to have this in the random provider. We are currently generating tokens for akamai CDN which require hex strings, and the only option I saw is to use the random_id resource, but that's a really insecure option since output is not sensitive :(

Hope you get soon a second approval for this PR.

Cheers!

**Comment by Socolin** [comment]
@bendbennett Hello, do you have any ETA or "Not before" date for this ?
I would like to know to take some decision about this, I might publish this in another provider so I can start using it, but I will not do this if it's close to be merged.

**Comment by bendbennett** [comment]
Hi @Socolin and @GerardSoleCa 👋 

Apologies for the delay. The focus of the team is currently elsewhere at the moment so it's hard to give an ETA.

**Comment by Socolin** [comment]
To anyone looking for this, I published this in a separate provider

https://registry.terraform.io/providers/Socolin/randombyte/latest

**Comment by bendbennett** [comment]
Hi @Socolin 👋 

Apologies for the delay in getting back to you. We have recently discussed your PR, and, if you're still amenable, we would like to move forward with the addition of the `random_bytes` resource.

The following suggestions came out of our discussion:

- Could `result_base64`, and `result_hex` be renamed `base64`, and `hex`, respectively?
- Could `id` be removed? Just to clarify, it is not used by practitioners, it is not required for the provider supported versions of Terraform (>= 0.12), and it is no longer needed for testing purposes (i.e., when the [terraform-plugin-testing](https://github.com/hashicorp/terraform-plugin-testing) module is used).
- With the addition of `random_bytes`, it would likely be valuable to update the documentation for both `random_id`, and `random_bytes` to provide some clarification around the use-case for using one or other of these two resources. The most obvious differentiator being when the output would be considered sensitive, and should not be displayed in CLI console output.

Thank you again for your contribution, and sorry it has taken a while for us to get back to you.

**Comment by bendbennett** [comment]
Hi @Socolin 👋 

I've incoporated all of your commits into [Add random_bytes resource](https://github.com/hashicorp/terraform-provider-random/pull/494#top).

If you'd rather update this PR with the suggestions in this [comment](https://github.com/hashicorp/terraform-provider-random/pull/272#issuecomment-1824621139), then that would be great. If not, then we'll move forward with [Add random_bytes resource](https://github.com/hashicorp/terraform-provider-random/pull/494#top).

Many thanks for all your work on this, and apologies for the delay.

**Comment by Socolin** [comment]
Hello,

I'm working on other project right now, you can move forward with your PR, Thanks :)

**Pull request closed without merging**
